File: Singleton 單例模式/Java/LazySingleton.java
public class LazySingleton {
    // 運行中途才創建實例
    private static LazySingleton instance;

    // 私有化建構子
    // 避免在外面被建構出來
    private LazySingleton() {}

    // 只能透過這邊取得實例
    // Double-Checked Locking
    public static LazySingleton getInstance() {
        // 避免拖累效能
        // 在外層先檢查一次
        if (instance == null) {
            // 限制只有一個thread可以進入
            // 避免多個thread同時進入
            // 而建立出多個實例
            synchronized (LazySingleton.class) {
                // 當實例尚未被建立出來
                if (instance == null) {
                    instance = new LazySingleton();
                }
            }
        }
        return instance;
    }
}

File: Singleton 單例模式/TypeScript/LazySingleton.ts
class LazySingleton {
  // 運行中途才創建實例
  private static instance: LazySingleton;

  // 私有化建構子
  // 避免在外面被建構出來
  private constructor() {}

  // 只能透過這邊取得實例
  // 跟Java不同的是
  // 由於nodejs只有單執行緒
  // 不用特別對執行緒做限制
  public static getInstance(): LazySingleton {
    if (this.instance === null) {
      this.instance = new LazySingleton();
    }
    return LazySingleton.instance;
  }
}
